components/kyma-environment-broker/internal/environmentscleanup/service.go

```diff
@@ -69,78 +71,112 @@ func (s *Service) Run() error {
 }
 
 func (s *Service) PerformCleanup() error {
-
-	staleShoots, err := s.getStaleShoots(s.LabelSelector)
+	runtimesToDelete, shootsToDelete, err := s.getStaleRuntimesByShoots(s.LabelSelector)
+	// runtimesToDelete, _, err := s.getStaleRuntimesByShoots(s.LabelSelector)
 	if err != nil {
 		s.logger.Error(fmt.Errorf("while getting stale shoots to delete: %w", err))
 		return err
 	}
 
-	runtimesToDelete := s.getRuntimes(staleShoots)
+	// return s.cleanupRuntimes(runtimesToDelete)
+	err = s.cleanupRuntimes(runtimesToDelete)
+	if err != nil {
+		s.logger.Error(fmt.Errorf("while cleaning runtimes: %w", err))
+		return err
+	}
 
-	s.logger.Infof("Runtimes to process: %+v\n", runtimesToDelete)
+	return s.cleanupShoots(shootsToDelete)
+}
 
-	if len(runtimesToDelete) == 0 {
+func (s *Service) cleanupRuntimes(runtimes []runtime) error {
+	s.logger.Infof("Runtimes to process: %+v\n", runtimes)
+
+	if len(runtimes) == 0 {
 		return nil
 	}
 
-	return s.cleanUp(runtimesToDelete)
+	return s.cleanUp(runtimes)
 }
 
-func (s *Service) getStaleShoots(labelSelector string) ([]unstructured.Unstructured, error) {
+func (s *Service) cleanupShoots(shoots []unstructured.Unstructured) error {
+	// do not log all shoots as previously - too much info
+	s.logger.Infof("Number of shoots to process: %+v\n", len(shoots))
+
+	if len(shoots) == 0 {
+		return nil
+	}
+
+	for _, shoot := range shoots {
+		err := s.gardenerService.Delete(context.Background(), shoot.GetName(), v1.DeleteOptions{})
+		if err != nil {
+			s.logger.Error(fmt.Errorf("while cleaning runtimes: %w", err))
+		}
+	}
+
+	return nil
+}
+
+func (s *Service) getStaleRuntimesByShoots(labelSelector string) ([]runtime, []unstructured.Unstructured, error) {
 	opts := v1.ListOptions{
 		LabelSelector: labelSelector,
 	}
 	shootList, err := s.gardenerService.List(context.Background(), opts)
 	if err != nil {
-		return []unstructured.Unstructured{}, fmt.Errorf("while listing Gardener shoots: %w", err)
+		return []runtime{}, []unstructured.Unstructured{}, fmt.Errorf("while listing Gardener shoots: %w", err)
 	}
 
+	var runtimes []runtime
 	var shoots []unstructured.Unstructured
 	for _, shoot := range shootList.Items {
 		shootCreationTimestamp := shoot.GetCreationTimestamp()
 		shootAge := time.Since(shootCreationTimestamp.Time)
 
-		if shootAge.Hours() >= s.MaxShootAge.Hours() {
-			log.Infof("Shoot %q is older than %f hours with age: %f hours", shoot.GetName(), s.MaxShootAge.Hours(), shootAge.Hours())
+		if shootAge.Hours() < s.MaxShootAge.Hours() {
+			continue
+		}
+
+		log.Infof("Shoot %q is older than %f hours with age: %f hours", shoot.GetName(), s.MaxShootAge.Hours(), shootAge.Hours())
+		staleRuntime, err := s.shootToRuntime(shoot)
+		if err != nil {
+			s.logger.Info("found a shoot without accounts references")
```

Comment:
or without runtime-id annotation...

Reply:
I changed it to "... without kcp references"